src/sources/exec.rs

```diff
@@ -0,0 +1,843 @@
+    Box::pin(tokio::spawn(async move {
+        info!("Start capturing {} command output.", stream);
+
+        let mut buffer: Vec<u8> = Vec::new();
+
+        let mut reader = reader.allow_read_until(shutdown.clone().map(|_| ()));
+
+        // Read one byte at a time so we don't block waiting for the buffer to
+        // fill up e.g. If a line filled up half the buffer we would not know about
+        // it until the buffer is returned. We could increase the read buffer if we are
+        // willing to accept some blocking waiting for it to fill up. I'm not sure if
+        // the underlying BufReader will read in larger chunks or not.
+        // Possibly could look into https://crates.io/crates/fixed-buffer-tokio but
+        // I don't know much about this library.
+        let mut read_buffer = [0_u8; 1];
+
+        // Not using the shutdown signal in this method so we need to shutdown
+        // in other methods to end the reading
+        while let Ok(bytes_read) = reader.read(&mut read_buffer).await {
+            if bytes_read == 0 {
+                info!("End of input reached, stop reading.");
+                break;
+            } else {
+                let read_byte = read_buffer[0];
+
+                // Could be enhanced to split 'lines' based on user defined
+                // delimiters in addition to newline.
+                if event_per_line && read_byte == b'\n' {
+                    if buffer.ends_with(&[b'\r']) {
+                        let _ = buffer.pop();
+                    }
+
+                    if let Some(buffer_string) = buffer_to_string(&mut buffer, false) {
+                        if sender.send((buffer_string, stream)).await.is_err() {
+                            break;
+                        }
+                    } else {
+                        info!("Invalid utf8, stop reading.");
+                        break;
+                    }
+                } else {
+                    buffer.push(read_byte);
+
+                    if buffer.len() == buf_size {
+                        if let Some(buffer_string) = buffer_to_string(&mut buffer, true) {
+                            if sender.send((buffer_string, stream)).await.is_err() {
+                                break;
+                            }
+                        } else {
+                            info!("Invalid utf8, stop reading.");
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        // Handle any left over buffer
+        if !buffer.is_empty() {
+            if let Some(buffer_string) = buffer_to_string(&mut buffer, true) {
+                let _ = sender.send((buffer_string, stream)).await;
+                if !buffer.is_empty() {
+                    info!("Invalid utf8, left in buffer.");
+                }
+            } else {
+                info!("Invalid utf8, left in buffer.");
+            }
```

Comment:
This would probably be better with a `match`:
```suggestion
            match buffer_to_string(&mut buffer, true) {
                Some(buffer_string) => {
                    let _ = sender.send((buffer_string, stream)).await;
                    if !buffer.is_empty() {
                        info!("Invalid utf8, left in buffer.");
                    }
                }
                None => info!("Invalid utf8, left in buffer."),
            }
```

Reply:
Thank you. I will use this if we add the method back in. It is removed for now.